local Shared = script.Parent.Parent.Shared

local ReplicatedStorage = game:GetService("ReplicatedStorage")
local RunService = game:GetService("RunService")

local Warn = require(Shared.Warn)
local StatsALL = require(script.Parent.Parent.Shared.Stats)
local ApplyMounts = require(script.Parent.Parent.Shared.ApplyMounts)
local RenderCache = require(script.Parent.ClientClock)
local ModelResolver = require(script.Parent.ModelResolver)
local Entity = require(Shared.Entity)
local Holder = require(Shared.Holder)
local Config = require(Shared.Config)
local Types = require(Shared.Types)
local Sender = require(script.Parent.Sender)
local receiver = require(script.Parent.Receiver)
local Stats = require(Shared.Stats).CLIENT
require(script.Parent.Player)
require(script.Parent.Receiver)

local idMap = Holder.idMap

local FrustumLookVector = vector.create(0, 0, -1)
local FrustumRightVector = vector.create(1, 0, 0)
local FrustumUpVector = vector.create(0, 1, 0)
local FrustumPosition = vector.create(0, 0, 0)
local TanHorizontal = 0
local TanVertical = 0

local function PrepareFrustumCheck()
	local cframe: CFrame = workspace.CurrentCamera.CFrame
	FrustumLookVector = cframe.LookVector :: any
	FrustumRightVector = cframe.RightVector :: any
	FrustumUpVector = cframe.UpVector :: any

	FrustumPosition = cframe.Position :: any

	local cam = workspace.CurrentCamera
	local halfFov = math.rad(cam.FieldOfView / 2)
	local aspectRatio = cam.ViewportSize.X / cam.ViewportSize.Y
	TanHorizontal = math.tan(halfFov) * aspectRatio
	TanVertical = math.tan(halfFov)
end

local function CheckPointInFrustum(pos: vector)
	local diff = pos - FrustumPosition

	local magnitude = vector.magnitude(diff)

	if magnitude == 0 then
		return true
	end

	local forwardDis = vector.dot(diff, FrustumLookVector)

	local forwardDot = forwardDis / magnitude
	if forwardDot < 0 then
		return false
	end
	if forwardDis < 1 then
		return true
	end
	local unit = diff / magnitude
	local rightDot = vector.dot(unit, FrustumRightVector)

	if math.abs(rightDot) > forwardDot * TanHorizontal then
		return false
	end

	local upDot = vector.dot(unit, FrustumUpVector)
	if math.abs(upDot) > forwardDot * TanVertical then
		return false
	end
	return true
end

local function CheckBroadPhaseInFrustum(center: vector, size: vector)
	if size == vector.zero then
		return CheckPointInFrustum(center)
	end
	local halfSize = size * 0.5
	local halfX = halfSize.x
	local halfY = halfSize.y
	local halfZ = halfSize.z
	local negX = -halfX
	local negY = -halfY
	local negZ = -halfZ

	return CheckPointInFrustum(center + vector.create(halfX, halfY, halfZ))
		or CheckPointInFrustum(center + vector.create(negX, halfY, halfZ))
		or CheckPointInFrustum(center + vector.create(halfX, negY, halfZ))
		or CheckPointInFrustum(center + vector.create(negX, negY, halfZ))
		or CheckPointInFrustum(center + vector.create(halfX, halfY, negZ))
		or CheckPointInFrustum(center + vector.create(negX, halfY, negZ))
		or CheckPointInFrustum(center + vector.create(halfX, negY, negZ))
		or CheckPointInFrustum(center + vector.create(negX, negY, negZ))
end
local Ticks = 4
local OutOfBroadPhaseTick = 0
local CurrentTick = 0
local TotalInterpolationTime = 0
local AmtInterpolations = 0

local CheckedThisFrame = 0
local TotalCulled = 0
local MovedThisFrame = 0

local function Interpolate(entity: Types.Entity)
	if not entity.interpolation or entity.entityConfig.CUSTOM_INTERPOLATION then
		return
	end
	local primaryPart = Entity.GetPrimaryPart(entity)

	if not primaryPart then
		return
	end
	if entity.isContextOwner then
		return
	end

	local currentEntityTick = entity._outofBroadPhaseTick
	if currentEntityTick and currentEntityTick ~= CurrentTick then
		return
	end

	local snapshot = entity.snapshot
	if not snapshot then
		Warn.medium("No snapshot for entity", entity)
		return
	end
	CheckedThisFrame += 1
	debug.profilebegin("Get Render CFrame")
	local targetTime = Entity.GetTargetRenderTime(entity)
	local targetCFrame = snapshot:GetAt(targetTime)
	debug.profileend()
	if not targetCFrame then
		return
	end
	local broadPhase: vector? = entity.broadPhase

	if broadPhase then
		local toCheck = targetCFrame :: any
		local lastCheckedCFrame = entity.lastCheckedCFrame
		debug.profilebegin("Prepare Frustum")
		local passed = CheckBroadPhaseInFrustum(toCheck.Position, broadPhase)
		debug.profileend()
		if not passed then
			if lastCheckedCFrame and lastCheckedCFrame.Position ~= toCheck.Position then
				debug.profilebegin("Check Last CFrame")
				local passLast = CheckBroadPhaseInFrustum(lastCheckedCFrame.Position :: any, broadPhase)
				debug.profileend()
				if not passLast then
					TotalCulled += 1
					return
				end
			else
				if not lastCheckedCFrame then
					entity.lastCheckedCFrame = toCheck
				end
				if not entity._outofBroadPhaseTick then
					entity._outofBroadPhaseTick = OutOfBroadPhaseTick
					OutOfBroadPhaseTick += 1
					if OutOfBroadPhaseTick % Ticks == 0 then
						OutOfBroadPhaseTick = 0
					end
				end
				TotalCulled += 1
				return
			end
		end
		entity.lastCheckedCFrame = nil
		entity._outofBroadPhaseTick = nil
	end
	debug.profilebegin("Set CFrame")
	if targetCFrame == targetCFrame then
		primaryPart.CFrame = targetCFrame
	end
	MovedThisFrame += 1
	debug.profileend()
end

local function updateEntityStats()
	if not StatsALL.SERVER._SHOULD_REPLICATE or Stats._PAUSE then
		return
	end
	local entities = {}
	for _, entity in idMap do
		local id = entity.id
		local config = entity.entityConfig
		local configName = config and config.NAME or "DEFAULT"
		local networkOwner = entity.networkOwner
		local isCharacter = entity._player ~= nil

		local lastRecvTime = entity.latestTime or 0
		local lastReplicatedTime = entity.lastReplicatedTime or 0

		local renderCache = entity._clientClock

		if not renderCache then
			local RenderCaches = entity.entityConfig.CLIENT_CLOCK
			if RenderCaches then
				renderCache = entity.isHalfTicked and RenderCaches.HALF or RenderCaches.NORMAL
			end
		end

		local targetRender = 0
		local averageLatency = 0
		local deviation = 0
		local bufferedTime = 0
		if renderCache then
			targetRender = renderCache:GetTargetRenderTime()
			local interpolationBuff = renderCache.buffer
			if interpolationBuff then
				averageLatency = interpolationBuff.averageLatency
				deviation = interpolationBuff.deviation
				bufferedTime = RenderCache._getBuffer(renderCache)
			end
		end
		local primaryPart = Entity.GetPrimaryPart(entity)
		local snapshot = entity.snapshot
		local currentPosition = primaryPart and primaryPart.Position
		local snapshots = {}
		if snapshot then
			for _, snap in snapshot.cache do
				table.insert(snapshots, { t = snap.t, value = snap.value, velocity = snap.velocity })
			end
			table.sort(snapshots, function(a, b)
				return a.t > b.t
			end)
		end
		local data = {
			id = id,
			networkOwner = networkOwner,
			isCharacter = isCharacter,
			config = configName,
			lastReceivedTime = lastRecvTime,
			lastReplicatedTime = lastReplicatedTime,
			targetTime = targetRender,
			averageLatency = averageLatency,
			deviation = deviation,
			bufferedTime = bufferedTime,
			currentPosition = currentPosition,
			snapshots = snapshots,
		}
		table.insert(entities, data)
	end
	Stats.CLIENT_ENTITIES = entities
end
Config._WaitForLock(function()
	RunService.Heartbeat:Connect(function(deltaTime: number)
		CurrentTick += 1
		if CurrentTick % Ticks == 0 then -- Runs every 4 frames at 60hz this is 15hz
			Stats.TOTAL_ENTITIES_CULLED = TotalCulled
			TotalCulled = 0
			CurrentTick = 0
		end
		RenderCache.UpdateAll(deltaTime)
		debug.profilebegin("Prepare Frustum Check")
		PrepareFrustumCheck()
		debug.profileend()
		local total = 0
		local startTime = os.clock()
		debug.profilebegin("Interpolate Entities")
		for _, entity in idMap do
			ModelResolver(entity)
			if entity.destroyed or entity.paused then
				continue
			end
			if entity.mountParentId then
				continue
			end
			total += 1
			Interpolate(entity)
		end
		debug.profileend()
		local endTime = os.clock()
		debug.profilebegin("Update Entity Stats")
		updateEntityStats()
		debug.profileend()
		TotalInterpolationTime += (endTime - startTime)
		AmtInterpolations += 1

		debug.profilebegin("Apply Mounts")
		ApplyMounts()
		debug.profileend()
		if not Stats._PAUSE then
			Stats.TOTAL_CLIENT_ENTITIES_CHECKED_THIS_FRAME = total
			Stats.ENTITIES_MOVED_THIS_FRAME = MovedThisFrame
			Stats.TOTAL_CLIENT_ENTITIES = total
			Stats.AVG_INTERPOLATION_TIME_MS = (TotalInterpolationTime / AmtInterpolations) * 1000
		end
		MovedThisFrame = 0
		CheckedThisFrame = 0
		if AmtInterpolations > 60 then
			TotalInterpolationTime = 0
			AmtInterpolations = 0
		end
	end)
	RunService.Heartbeat:Connect(Sender.Update)

	Shared.Remotes.ClientLoaded:FireServer()
end)

return nil
